### PR TITLE
External idp impl

### PR DESCRIPTION
**Pull request by flo-renaud**
- External IdP schema: use octetstring syntax for client secret
- External Identity Provider plugin implementation
- External Idp: add XMLRPC tests

### CODE REVIEW

File: ipaserver/plugins/idp.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+
+@register()
+class idp_add(LDAPCreate):
+    __doc__ = _('Add a new Identity Provider server.')
+    msg_summary = _('Added Identity Provider server "%(value)s"')
+
+    # List of pre-populated idp endpoints
+    # key = provider, value = tuple of (auth-uri, token-uri)
+    idp_providers = {
+        u'google': (
```
Comment (abbra) [inline ipaserver/plugins/idp.py:184]: you don't need to use `u'...'` with Python 3 anymore, just use normal strings here and in other places, please.

File: ipaserver/plugins/idp.py [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    )
+
+    permission_filter_objectclasses = ['ipaidp']
+    managed_permissions = {
+        'System: Add External IdP server': {
+            'ipapermright': {'add'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
```
Comment (abbra) [inline ipaserver/plugins/idp.py:131]: Please change `Idp` to `IdP` here.

File: ipaserver/plugins/idp.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+            'ipapermright': {'read', 'search', 'compare'},
+            'ipapermdefaultattr': {
+                'cn', 'objectclass', 'ipaidpauthendpoint',
+                'ipaidptokenendpoint', 'ipaidpclientid',
+                'ipaidpscope'
+            },
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
```
Comment (abbra) [inline ipaserver/plugins/idp.py:143]: Please change `Idp` to `IdP` here.

File: ipaserver/plugins/idp.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        },
+        'System: Modify External IdP server': {
+            'ipapermright': {'write'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermdefaultattr': {
+                'cn', 'objectclass', 'ipaidpauthendpoint',
+                'ipaidptokenendpoint', 'ipaidpclientid',
+                'ipaidpscope', 'ipaidpclientsecret',
+            },
+            'default_privileges': {'External Idp server Administrators'}
```
Comment (abbra) [inline ipaserver/plugins/idp.py:153]: Please change `Idp` to `IdP` here.

File: ipaserver/plugins/idp.py [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+        },
+        'System: Delete External IdP server': {
+            'ipapermright': {'delete'},
+            'ipapermlocation': DN(container_dn, api.env.basedn),
+            'ipapermtargetfilter': {
+                '(objectclass=ipaidp)'},
+            'default_privileges': {'External Idp server Administrators'}
```
Comment (abbra) [inline ipaserver/plugins/idp.py:160]: Please change `Idp` to `IdP` here.

File: ipatests/test_xmlrpc/test_idp_plugin.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    def test_create_with_invalid_provider(self):
+        """ Creation with invalid --provider parameter """
+        idp_with_provider = IdpTracker(
+            u'idp_with_provider', ipaidpprovider=u'fake',
+            ipaidpclientid=u"idpclient1")
+        idp_with_provider.track_create()
+        command = idp_with_provider.make_create_command()
+        with raises_exact(errors.ValidationError(
+            name='provider',
+            error=u"must be one of 'google', 'github', 'microsoft-common', "
```
Comment (abbra) [inline ipatests/test_xmlrpc/test_idp_plugin.py:177]: Same here -- no need to use `u"..."` for strings anymore

**Review by abbra — CHANGES_REQUESTED** [review]
Thank you for the PR. It looks fine in general, couple comments inline. There are many places where you have used `u"..."` for unicode strings but this is not really needed with Python 3 anymore, could you please remove all the prefixes?
Also should we add 'admins' group to the default privilege?

**Comment by flo-renaud** [comment]
Hi @abbra 
thanks for the review. I removed the unicode prefix (hope I didn't forget it anywhere) and fixed the uppercase in IdP.

> Also should we add 'admins' group to the default privilege?

By default admins can manage any entry thanks to the following ACI on the basedn:
```
aci: (targetattr != "userPassword || krbPrincipalKey || sambaLMPassword || sam
 baNTPassword || passwordHistory || krbMKey || krbPrincipalName || krbCanonica
 lName || krbPwdHistory || krbLastPwdChange || krbExtraData || krbLastSuccessf
 ulAuth || krbLastFailedAuth || ipaUniqueId || memberOf || enrolledBy || ipaNT
 Hash || ipaProtectedOperation")(version 3.0; acl "Admin can manage any entry"
 ; allow (all) groupdn = "ldap:///cn=admins,cn=groups,cn=accounts,dc=ipa,dc=te
 st";)

```

**Comment by abbra** [comment]
Thanks. I think there is one place with `u'..'` left in tests. API.txt file has `u'..'` as well but those are generated and kept this way so they are OK.

**Comment by flo-renaud** [comment]
I forgot to re-run makeaci and makeapi, it may fix some of the u'' issues. I'm on it...

**Comment by abbra** [comment]
Right, it was not in the new code -- only in the context lines.

**Comment by abbra** [comment]
Merged, thanks!

**Pull request merged**